aten/src/ATen/DynamicLibrary.cpp

```diff
@@ -46,11 +46,31 @@ DynamicLibrary::~DynamicLibrary() {
 
 DynamicLibrary::DynamicLibrary(const char* name) {
   // NOLINTNEXTLINE(hicpp-signed-bitwise)
-  HMODULE theModule = LoadLibraryA(name);
+  HMODULE theModule;
+  bool reload = true;
+  if (GetProcAddress(GetModuleHandle("KERNEL32.DLL"), "AddDllDirectory") != NULL) {
+    theModule = LoadLibraryExA(
+        name,
+        NULL,
+        LOAD_LIBRARY_SEARCH_DEFAULT_DIRS);
+    if (theModule != NULL || (GetLastError() != ERROR_MOD_NOT_FOUND)) {
+      reload = false;
+    }
+  }
+
+  if (reload) {
+    theModule = LoadLibraryA(name);
+  }
+
```

Comment:
Why `theModule != NULL` is not a sufficient check? Moreover, why simply not do something like:
`theModule = has_LOAD_LIBRARY_SEARCH_DEFAULT_DIRS() ? LoadLibraryEx(...) : LoadLibrary(...);`
```suggestion
  HMODULE theModule = NULL;
  if (GetProcAddress(GetModuleHandle("KERNEL32.DLL"), "AddDllDirectory") != NULL) {
    theModule = LoadLibraryExA(
        name,
        NULL,
        LOAD_LIBRARY_SEARCH_DEFAULT_DIRS);
  }

  if (theModule == NULL) {
    theModule = LoadLibraryA(name);
  }

```

Reply:
Because the main differences between LoadLibrary and LoadLibraryEx is that the search order and the directories being searched. So only when we get `ERROR_MOD_NOT_FOUND`, it is meaningful to try LoadLibrary afterwards.

Reply:
I still can't imagine a valid PyTorch installation, where `LoadLibraryEx` would fail but `LoadLibrary` will succeed. 
Alternatively, if we will fall-back to `LoadLibrary` if `LoadLibraryEx` would fail, shouldn't we always call `LoadLibrary` (because, it would not ignore paths specified via `AddDllDirectory`, would it?)

Or, why not do something like that:
```
 DWORD loadFlags = 0;
 if (hasAddDllDirectory) loadFlags = LOAD_LIBRARY_SEARCH_DEFAULT_DIRS;
 theModule = loadLibraryExA(name, NULL, loadFlags);
```

Reply:
> I still can't imagine a valid PyTorch installation, where `LoadLibraryEx` would fail but `LoadLibrary` will succeed.

LoadLibraryExA with LOAD_LIBRARY_SEARCH_DEFAULT_DIRS doesn't considers directories in PATH.


> Alternatively, if we will fall-back to `LoadLibrary` if `LoadLibraryEx` would fail, shouldn't we always call `LoadLibrary` (because, it would not ignore paths specified via `AddDllDirectory`, would it?)

Yes, it is partially the reason. Also, the search order is more reasonable because SYSTEM dir is searched in the last step.